src/consensus/produce_block.rs

```diff
@@ -126,45 +112,46 @@ where
             parent_slot,
         );
 
-        let mut sleep_duration = DELTA_BLOCK;
-        for i in 0.. {
-            let slice_index = if i == 0 {
-                Either::Left((parent_block_id, 0))
+        let mut duration_left = DELTA_BLOCK;
+        for slice_index in 0.. {
+            let parent = if slice_index == 0 {
+                Some(parent_block_id)
             } else {
-                Either::Right(NonZeroUsize::new(i).unwrap())
+                None
             };
-            let (slice, cont_prod) =
-                produce_slice(&self.txs_receiver, slot, slice_index, sleep_duration).await;
-            if i == 0 {
-                assert!(slice.parent.is_some());
-            }
+            let (payload, maybe_duration) =
+                produce_slice_payload(&self.txs_receiver, parent, duration_left).await;
 
-            let is_last = slice.is_last;
-            let maybe_block_hash = self.shred_and_disseminate(slice).await?;
-            match cont_prod {
-                Continue::Stop => {
-                    assert!(is_last);
-                    let block_hash = maybe_block_hash.unwrap();
-                    return Ok((slot, block_hash));
-                }
-                Continue::Continue { left } => {
-                    assert!(!is_last);
-                    assert!(maybe_block_hash.is_none());
-                    sleep_duration = left;
-                }
+            match self
+                .shred_and_disseminate(slice_index, slot, payload, maybe_duration)
+                .await?
+            {
+                Either::Left(block_hash) => return Ok((slot, block_hash)),
+                Either::Right(duration) => duration_left = duration,
             }
         }
         unreachable!()
     }
 
-    /// Shreds and disseminates the slice.
+    /// Shreds and disseminates the slice payload.
     ///
-    /// Returns Ok(Some(Hash)) if slice.is_last == true.
-    /// Returns Ok(None) if slice.is_last == false.
-    async fn shred_and_disseminate(&self, slice: Slice) -> Result<Option<Hash>> {
+    /// Returns Ok(Either::Left(hash of the block)) if this is the last slice.
+    /// Returns Ok(Either::Right(duration left in block)) if this is not the last slice.
+    async fn shred_and_disseminate(
+        &self,
+        slice_index: usize,
+        slot: Slot,
+        payload: SlicePayload,
+        maybe_duration: Option<Duration>,
+    ) -> Result<Either<Hash, Duration>> {
+        let is_last = slice_index == MAX_SLICES_PER_BLOCK - 1 || maybe_duration.is_none();
+        let header = SliceHeader {
+            is_last,
+            slot,
+            slice_index,
+        };
+        let slice = Slice::from_parts(header, payload, None);
```

Comment:
why do we construct the slice from payload here now, instead of passing it to this function? seems slightly more complicated if anything (more arguments and we have to determine `is_last` in two places)

Reply:
You are right, my original thinking was that I can reduce code duplication however, then I had to check for `is_last` in the other place.

I think I will still pass a SliceHeader instead of Slice to this function.  And we could then change `shred()` to maybe take the header and payload separately.  No need to construct the slice just to deconstruct it right away.